Add "clock_tag" field to update operations in the WAL and to the internal gRPC API

ffuugoo opened the pull request:
This PR adds a `clock_tag` field to update operations stored in WAL and to the internal gRPC API.

This field will be used (in follow-up PRs) to optimize shard transfer: e.g., search relevant position in WAL on a "good" node and only transfer part of WAL that is missing on "dead" node.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [ ] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

generall commented on lib/collection/src/collection/point_ops.rs:
```diff
@@ -103,7 +103,11 @@ impl Collection {
                 .map(move |(replica_set, operation)| {
-                    replica_set.update_with_consistency(operation, wait, ordering)
+                    replica_set.update_with_consistency(
+                        operation.into(), // TODO: Assign tag!
```
I would prefer to use explicit conversion `WithMeta::from(operation)`, as in this case navigation and code usage search is much simpler. (yes, even with rust_analyser)

generall commented on lib/collection/src/operations/mod.rs:
```diff
@@ -38,9 +38,68 @@ pub enum FieldIndexOperations {
+    pub metadata: Option<OperationMeta>,
+}
+
+impl WithMeta {
```
`WithMeta` looks more like a trait name. Maybe `OperationWithMeta`? 
Also `metadata` in general might be mixed up with payload. Not sure what would be a better alternative, though

generall commented on lib/collection/src/operations/mod.rs:
```diff
@@ -179,18 +238,188 @@ impl CollectionUpdateOperations {
+        #[test]
+        fn tagged_operation_serde(tagged in any::<WithMeta>()) {
+            // Assert that `TaggedOperation` can be serialized
+            let input = serde_json::to_string(&tagged).unwrap();
```
WAL uses `serde_cbor`, maybe worth testing with it as well?

generall reviewed (changes requested):
Renaming + minor stuff mentioned in comments, the rest LGTM

ffuugoo commented:
@generall Rebased on recent `dev`, added `timestamp` field to internal point update gRPC API (from #3426), renamed `WithMeta` into `OperationWithTimestamp` and `OperationMeta` into `OperationTimestamp`.

ffuugoo commented:
Rebased on `dev`. (Had to squash everything to simplify rebase, sorry.)

ffuugoo commented:
There are a few `TODO`s scattered around the code, I'll fix them up in a few hours.

timvisee reviewed (approved):
I don't like the TODOs, but I don't see a better way of doing things (assuming we cannot fix them in this PR).

ffuugoo commented:
> I don't like the TODOs, but I don't see a better way of doing things (assuming we cannot fix them in this PR).

Some are left there intentionally to be fixed in the follow-up PRs, so I'll fix today before merging.

ffuugoo commented on lib/collection/src/shards/forward_proxy_shard.rs:
```diff
@@ -56,12 +58,13 @@ impl ForwardProxyShard {
             self.remote_shard
                 .update(
-                    CollectionUpdateOperations::FieldIndexOperation(
+                    // TODO: Assign clock tag!? 🤔
```
Well, yes, but I'd still leave it for now until we implement all parts completely. Maybe something else comes up and would be able/have to assign the tag. Having it marked with `TODO` makes it a bit easier to find/harder to overlook.

timvisee commented:
Great! CI is green now. I'll give you the honor to merge.
